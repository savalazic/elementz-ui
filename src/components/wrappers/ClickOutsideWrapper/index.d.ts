export { default, ClickOutsideWrapperProps } from './ClickOutsideWrapper';
