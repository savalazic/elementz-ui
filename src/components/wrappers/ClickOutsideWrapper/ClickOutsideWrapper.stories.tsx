import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { infoConfig } from '../../../utils/info-config';

import { ClickOutsideWrapper } from './ClickOutsideWrapper';

const stories = storiesOf('Wrappers|ClickOutsideWrapper', module);

stories.addParameters(infoConfig);

stories.add('default', () => {
  return (
    <ClickOutsideWrapper onClickOutside={action('click outside wrapper')}>
      <div
        style={{
          padding: 30,
          backgroundColor: '#ddd',
        }}
      >
        <p style={{ textAlign: 'center' }}>Click outside this area</p>
      </div>
    </ClickOutsideWrapper>
  );
});
