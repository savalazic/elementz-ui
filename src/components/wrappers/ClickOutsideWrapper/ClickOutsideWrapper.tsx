import React, { PureComponent } from 'react';

export interface ClickOutsideWrapperProps {
  /** function callback */
  onClickOutside: (event: any) => void;
}

export class ClickOutsideWrapper extends PureComponent<
  ClickOutsideWrapperProps
> {
  private wrapperRef: React.RefObject<any> = React.createRef();

  public handleClickOutside = (event: any) => {
    if (this.wrapperRef && !this.wrapperRef.current.contains(event.target)) {
      this.props.onClickOutside(event);
    }
  };

  public componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
  }

  public componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
  }

  public render() {
    const { children } = this.props;

    return <div ref={this.wrapperRef}>{children}</div>;
  }
}
