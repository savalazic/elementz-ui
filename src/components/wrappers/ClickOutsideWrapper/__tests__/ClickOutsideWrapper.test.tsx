import 'jest-dom/extend-expect';
import React from 'react';
import { fireEvent, render } from 'react-testing-library';

import { ClickOutsideWrapper } from '../ClickOutsideWrapper';

describe('ClickOutsideWrapper', () => {
  const handleClick = jest.fn();

  it('should match default snapshot', () => {
    const { container } = render(
      <ClickOutsideWrapper onClickOutside={handleClick} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should not call handleClick when user clicks inside wrapper', () => {
    const { getByTestId } = render(
      <ClickOutsideWrapper onClickOutside={handleClick}>
        <div data-testid="area">
          <p>Click outside this area</p>
        </div>
      </ClickOutsideWrapper>,
    );
    fireEvent.click(getByTestId('area'));
    fireEvent.click(getByTestId('area'));
    fireEvent.click(getByTestId('area'));
    expect(handleClick).toHaveBeenCalledTimes(0);
  });

  it('should call handleClick when user clicks outside wrapper', () => {
    const { getByTestId } = render(
      <div>
        <ClickOutsideWrapper onClickOutside={handleClick}>
          <div>
            <p>Click outside this area</p>
          </div>
        </ClickOutsideWrapper>
        <div data-testid="area">
          <p>Click here</p>
        </div>
      </div>,
    );
    fireEvent.mouseDown(getByTestId('area'));
    fireEvent.mouseDown(getByTestId('area'));
    fireEvent.mouseDown(getByTestId('area'));
    expect(handleClick).toHaveBeenCalledTimes(3);
  });
});
