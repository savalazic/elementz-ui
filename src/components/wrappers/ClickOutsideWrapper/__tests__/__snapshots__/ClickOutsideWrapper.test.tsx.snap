// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickOutsideWrapper should match default snapshot 1`] = `
<div>
  <div />
</div>
`;
